src/storage/lock_manager/deadlock.rs

```diff
@@ -5,36 +5,45 @@ use super::config::Config;
 use super::metrics::*;
 use super::waiter_manager::Scheduler as WaiterMgrScheduler;
 use super::{Error, Lock, Result};
+use crate::raftstore::coprocessor::{Coprocessor, CoprocessorHost, ObserverContext, RoleObserver};
 use crate::server::resolve::StoreAddrResolver;
 use futures::{Future, Sink, Stream};
 use grpcio::{
     self, DuplexSink, RequestStream, RpcContext, RpcStatus, RpcStatusCode, UnarySink, WriteFlags,
 };
 use kvproto::deadlock::*;
+use kvproto::metapb::Region;
 use pd_client::{RpcClient, INVALID_ID};
+use raft::StateRole;
 use std::cell::RefCell;
 use std::fmt::{self, Display, Formatter};
 use std::rc::Rc;
 use std::sync::Arc;
-use std::time::{Duration, Instant};
 use tikv_util::collections::{HashMap, HashSet};
 use tikv_util::future::paired_future_callback;
 use tikv_util::security::SecurityManager;
-use tikv_util::time;
-use tikv_util::worker::{FutureRunnable, FutureScheduler, Stopped};
+use tikv_util::time::{Duration, Instant};
+use tikv_util::worker::{FutureRunnable, FutureScheduler, FutureWorker, Stopped};
 use tokio_core::reactor::Handle;
-use tokio_timer::Interval;
+
+/// The leader of the region containing the LEADER_KEY is the leader of deadlock detector.
+const LEADER_KEY: &[u8] = b"";
+
+/// Returns true if the region containing the LEADER_KEY.
+fn is_leader_region(region: &'_ Region) -> bool {
+    region.get_start_key() <= LEADER_KEY && LEADER_KEY < region.get_end_key()
```

Comment:
When talking about end_key, we use an empty key to represent the maximum key (∞). So when you want to check if a key is less than a region's end_key, you checks `region.get_end_key().is_empty() || key < region.get_end_key()`.
```suggestion
    region.get_start_key() <= LEADER_KEY && (region.get_end_key().is_empty() || LEADER_KEY < region.get_end_key())
```